File: k6/silver.js
/**
 * SILVER -- Scale-Out: ramp to 200 concurrent users
 *
 * Pre-requisites:  docker compose up -d --scale app=2
 * Run:             k6 run k6/silver.js
 * Target:          p95 < 3s, error rate < 5%
 */
import http from "k6/http";
import { check, sleep } from "k6";
import { Rate } from "k6/metrics";

const errorRate = new Rate("error_rate");

export const options = {
  stages: [
    { duration: "30s", target: 200 }, // ramp up
    { duration: "2m", target: 200 }, // hold
    { duration: "15s", target: 0 }, // ramp down
  ],
  thresholds: {
    http_req_duration: ["p(95)<3000"],
    error_rate: ["rate<0.05"],
  },
};

const BASE = "http://localhost";

export default function () {
  const roll = Math.random();
  if (roll < 0.5) {
    const res = http.get(`${BASE}/api/products`);
    errorRate.add(!check(res, { "products 200": (r) => r.status === 200 }));
  } else if (roll < 0.75) {
    const res = http.get(`${BASE}/api/inventory`);
    errorRate.add(!check(res, { "inventory 200": (r) => r.status === 200 }));
  } else {
    const res = http.post(
      `${BASE}/api/sales`,
      JSON.stringify({
        cashier_id: Math.ceil(Math.random() * 5),
        items: [{ product_id: Math.ceil(Math.random() * 50), quantity: 1 }],
      }),
      { headers: { "Content-Type": "application/json" } },
    );
    errorRate.add(!check(res, { "sale 201": (r) => r.status === 201 }));
  }
  sleep(1);
}

File: locust/silver.py
"""
SILVER — Scale-Out: ramp to 200 concurrent users
══════════════════════════════════════════════════
Pre-requisites:
    docker compose up -d --scale app=2

Run headless:
    locust -f locust/silver.py \
           --headless --host http://localhost \
           --html reports/silver_report.html

The LoadTestShape below controls the ramp automatically.
No --users flag needed — the shape overrides it.

Target: p95 < 3 s, failure rate < 5 %
"""

import random

from locust import HttpUser, LoadTestShape, constant, task


class SilverShape(LoadTestShape):
    """
    Drives the user count over time.
    tick() is called every second — return (target_users, spawn_rate) or None to stop.

    Timeline:
        0 s  → 30 s  : ramp up to 200 users  (10 new users/sec)
        30 s → 150 s : hold at 200 users      (2 minutes of sustained load)
        150 s → 165 s: ramp down to 0         (15 s cool-down)
    """

    stages = [
        {"until": 30, "users": 200, "spawn_rate": 10},
        {"until": 150, "users": 200, "spawn_rate": 10},
        {"until": 165, "users": 0, "spawn_rate": 20},
    ]

    def tick(self):
        t = self.get_run_time()
        for stage in self.stages:
            if t <= stage["until"]:
                return stage["users"], stage["spawn_rate"]
        return None  # returning None stops the test cleanly


class CashierUser(HttpUser):
    wait_time = constant(1)
    host = "http://localhost"

    @task(2)
    def browse_products(self):
        with self.client.get(
            "/api/products", name="GET /api/products", catch_response=True
        ) as resp:
            if resp.status_code != 200:
                resp.failure(f"Expected 200, got {resp.status_code}")

    @task(1)
    def check_inventory(self):
        with self.client.get(
            "/api/inventory", name="GET /api/inventory", catch_response=True
        ) as resp:
            if resp.status_code != 200:
                resp.failure(f"Expected 200, got {resp.status_code}")

    @task(1)
    def create_sale(self):
        with self.client.post(
            "/api/sales",
            json={
                "cashier_id": random.randint(1, 5),
                "items": [{"product_id": random.randint(1, 50), "quantity": 1}],
            },
            name="POST /api/sales",
            catch_response=True,
        ) as resp:
            if resp.status_code != 201:
                resp.failure(f"Expected 201, got {resp.status_code}")
